src/foam/nanos/session/Session.js

```diff
@@ -185,25 +189,31 @@ foam.CLASS({
       `
     },
     {
-      name: 'validRemoteHost',
-      type: 'Boolean',
+      name: 'validateRemoteHost',
       args: [
         {
-          name: 'remoteHost', type: 'String'
+          name: 'x',
+          type: 'Context'
         }
       ],
+      javaThrows: ['foam.core.ValidationException'],
       javaCode: `
-        if ( SafetyUtil.isEmpty(getRemoteHost()) || SafetyUtil.equals(getRemoteHost(), remoteHost) ) {
-          return true;
-        }
-
-        for ( String host : getRemoteHostWhiteList() ) {
-          if ( SafetyUtil.equals(host, remoteHost) ) {
-            return true;
+      String remoteIp = foam.net.IPSupport.instance().getRemoteIp(x);
+      if ( remoteIp == null ||
+           SafetyUtil.isEmpty(getRemoteHost()) ||
+           SafetyUtil.equals(getRemoteHost(), remoteIp) ) {
+        return;
+      }
+      for ( foam.net.CIDR cidr : getCidrWhiteList() ) {
+        try {
+          if ( cidr.inRange(x, remoteIp) ) {
+            return;
           }
+        } catch (java.net.UnknownHostException e) {
+          ((foam.nanos.logger.Logger) x.get("logger")).warning(this.getClass().getSimpleName(), "validateRemoteHost", remoteIp, e.getMessage());
         }
-
-        return false;
+      }
+      throw new foam.core.ValidationException("Restricted IP");
```

Comment:
Getting "Access denied" when login as admin@nanopay.net.
```
2021-02-12T07:08:38.421-0500,qtp1189257362-1848,DEBUG,[Service],auth,Unauthenticated session,returning skeleton exception,[localhost],foam.nanos.auth.AccessDeniedException,Access denied
	at foam.nanos.auth.UserAndGroupAuthService.loginHelper(UserAndGroupAuthService.java:335)
	at foam.nanos.auth.UserAndGroupAuthService.login(UserAndGroupAuthService.java:390)
	at net.nanopay.auth.NanopayUserAndGroupAuthService.login_(NanopayUserAndGroupAuthService.java:228)
	at net.nanopay.auth.NanopayUserAndGroupAuthService.login(NanopayUserAndGroupAuthService.java:191)
	at foam.nanos.auth.ProxyAuthService.login(ProxyAuthService.java:337)
	at foam.nanos.auth.ProxyAuthService.login(ProxyAuthService.java:337)
	at foam.nanos.auth.ProxyAuthService.login(ProxyAuthService.java:337)
	at foam.nanos.auth.ProxyAuthService.login(ProxyAuthService.java:337)
	at foam.nanos.auth.PasswordExpiryAuthService.login(PasswordExpiryAuthService.java:120)
	at foam.nanos.auth.ProxyAuthService.login(ProxyAuthService.java:337)
	at net.nanopay.security.auth.LoginAttemptAuthService.login_(LoginAttemptAuthService.java:546)
	at net.nanopay.security.auth.LoginAttemptAuthService.login(LoginAttemptAuthService.java:503)
	at foam.nanos.auth.ProxyAuthService.login(ProxyAuthService.java:337)
	at net.nanopay.security.auth.IPLoggingAuthService.login(IPLoggingAuthService.java:92)
	at foam.nanos.auth.ProxyAuthService.login(ProxyAuthService.java:337)
	at foam.nanos.auth.ProxyAuthService.login(ProxyAuthService.java:337)
	at foam.nanos.auth.AuthServiceSkeleton.send(AuthServiceSkeleton.java:104)
	at foam.box.SessionServerBox.send(SessionServerBox.java:185)
	at foam.nanos.http.ServiceWebAgent.execute(ServiceWebAgent.java:129)
	at foam.nanos.http.NanoRouter.service(NanoRouter.java:127)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:790)
	at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:763)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1631)
	at org.eclipse.jetty.websocket.server.WebSocketUpgradeFilter.doFilter(WebSocketUpgradeFilter.java:226)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1618)
	at net.nanopay.security.csp.CSPFilter.doFilter(CSPFilter.java:24)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1618)
	at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:549)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:233)
	at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1369)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:188)
	at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:489)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:186)
	at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1284)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:127)
	at org.eclipse.jetty.server.Server.handle(Server.java:501)
	at org.eclipse.jetty.server.HttpChannel.lambda$handle$1(HttpChannel.java:383)
	at org.eclipse.jetty.server.HttpChannel$$Lambda$100/0000000000000000.dispatch(Unknown Source)
	at org.eclipse.jetty.server.HttpChannel.dispatch(HttpChannel.java:556)
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:375)
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:272)
	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:311)
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:103)
	at org.eclipse.jetty.io.ChannelEndPoint$1.run(ChannelEndPoint.java:104)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.runTask(EatWhatYouKill.java:336)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:313)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.tryProduce(EatWhatYouKill.java:171)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.run(EatWhatYouKill.java:129)
	at org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:375)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:806)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:938)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: [localhost],foam.core.ValidationException,Restricted IP
	at foam.nanos.auth.Group.validateCidrWhiteList(Group.java:1426)
	at foam.nanos.auth.UserAndGroupAuthService.loginHelper(UserAndGroupAuthService.java:333)
	... 52 more
```

Reply:
Are you saying you get this with no configuration?

Reply:
Updated.